web/packages/teleport/src/components/hooks/useInfiniteScroll.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  AgentResponse,
+  UnifiedResource,
+  AgentFilter,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ */
+export function useInfiniteScroll<T extends UnifiedResource>({
+  fetchFunc,
+  clusterId,
+  params,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const { attempt, setAttempt } = useAttempt('processing');
+
+  const [fetchedData, setFetchedData] = useState<AgentResponse<T>>({
+    agents: [],
+    startKey: '',
+    totalCount: 0,
+  });
+
+  const fetch = async () => {
```

Comment:
```suggestion
  const fetchInit = async () => {
```

Reply:
I'll go with "fetchInitial", since "fetchInit" sounds as if "init" was a verb, and it's a bit confusing (would it actually fetch, or just initialize fetching?)